doc/reference/customdata.md

```diff
@@ -0,0 +1,105 @@
+# Custom Data Management
+
+The `Custom Data` feature implements a storage for 'any' data, which can be used by other components for data persistence.
+
+Because the schema of the data being persisted varies from components, a `CustomDataKind` must be defined to distinguish the data.
+
+## CustomDataKind
+
+The YAML example below defines a CustomDataKind:
+
+```yaml
+name: kind1
+idField: name
+jsonSchema:
+  type: object
+  properties:
+    name:
+	  type: string
+	  required: true
+```
+
+The `name` field is required, it is the kind name of custom data items of this kind.
+The `idField` is optional, and its default value is `name`, the value of this field of a data item is used as its identifier.
+The `jsonSchema` is optional, if provided, data items of this kind will be validated against this [JSON Schema](http://json-schema.org/).
+
+## CustomData
+
+CustomData is a map, the keys of this map must be strings while the values can be any valid JSON values, but when value is also a map, its keys must be strings too.
+
```

Comment:
Probably very rare edge case but what about map containing slice/array of maps? In
`map[string][[]map[string]sometype]` I guess `sometype` can also only be `string` but not `slice` or `map`?

Is it correct to say:
```suggestion
In other words, nested CustomData objects can have at maximum the depth of two levels.
```

Reply:
No, there's no limitation for the depth.
I mean if there are nested maps (either as a value or as an array item), the keys of the map must be strings.

Reply:
Ah Ok I get it now.